homeassistant/components/ws980wifi/sensor.py

```diff
@@ -0,0 +1,322 @@
+    for sensor_property in config[CONF_MONITORED_CONDITIONS]:
+        sensors.append(WeatherSensor(name, SENSOR_PROPERTIES[sensor_property]))
+
+    weather = WeatherData(hass, sensors, config)
+    await weather.fetching_data()
+    async_add_entities(sensors)
+
+
+class WeatherSensor(Entity):
+    """Sensor class for all sensor type of this weatherstation."""
+
+    def __init__(self, name, sensor_property):
+        """Initialize the sensor."""
+        self.client_name = name
+        self._state = None
+        self._name = sensor_property[0]
+        self._unit_of_measurement = sensor_property[1]
+        self._device_class = sensor_property[2]
+        self._hex_index = int(sensor_property[3])
+        self._hex_length = int(sensor_property[4])
+        self._decimal_place = int(sensor_property[5])
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return f"{self.client_name} {self._name}"
+
+    @property
+    def state(self):
+        """Return the state of the device."""
+        return self._state
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return {ATTR_ATTRIBUTION: ATTRIBUTION}
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement of this entity, if any."""
+        return self._unit_of_measurement
+
+    @property
+    def device_class(self):
+        """Return the device class of this entity, if any."""
+        return self._device_class
+
+    @property
+    def hex_index(self):
+        """Return the hex index."""
+        return self._hex_index
+
+    @property
+    def hex_length(self):
+        """Return the hex length."""
+        return self._hex_length
+
+    @property
+    def decimal_place(self):
+        """Return the decimal."""
+        return self._decimal_place
+
+
+class WeatherData(Entity):
+    """Get the latest data and updates the states."""
+
+    def __init__(self, hass, sensors, config):
+        """Initialize the data object."""
+        self.hass = hass
+        self.sensors = sensors
+        self._config = {
+            CONF_HOST: config.get(CONF_HOST),
+            CONF_PORT: config.get(CONF_PORT),
+            CONF_TIMEOUT: config.get(CONF_TIMEOUT),
+            CONF_PAYLOAD: b"\xff\xff\x0b\x00\x06\x04\x04\x19",
+            CONF_BUFFER_SIZE: DEFAULT_BUFFER_SIZE,
+            CONF_SCAN_INTERVAL: DEFAULT_SCAN_INTERVAL,
+        }
+
+    async def fetching_data(self, *_):
+        """Get the data from weather station."""
+        _LOGGER.debug("Updating sensor values from weather station")
+
+        def try_again():
+            """Retry in few seconds."""
+            seconds = self._config[CONF_SCAN_INTERVAL] * 2
+            _LOGGER.error("Retrying in %i seconds", seconds)
+            async_call_later(self.hass, seconds, self.fetching_data)
+
+        data = None
+
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
```

Comment:
Please extract the device specific interface code and data parsing to a standalone library published on PyPI.

https://developers.home-assistant.io/docs/creating_component_code_review#4-communication-with-devicesservices